### PR TITLE
Improvements for associative_scan - Reverse feature

### PR DESCRIPTION
**Pull request by bohnstingl**
This is part of a series of PRs to improve the functionality of the `associatve_scan` functionality. This specific PR introduces a `reverse` flag to the `associative_scan` to establish a similar interface as for `jax.associative_scan`. This PR has been derived from https://github.com/pytorch/pytorch/pull/129307.

@ydwu4 @Chillee @zou3519

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_higher_order_ops/associative_scan.py [hunk 1]
```diff
@@ -77,11 +79,14 @@ def add(x: torch.Tensor, y: torch.Tensor):
...
         with _set_compilation_env(), torch._dynamo.utils.disable_cache_limit():
             return torch.compile(associative_scan, fullgraph=True)(
-                combine_fn, input, dim
+                combine_fn, input, dim, reverse=reverse
             )
 
     leaves, spec = pytree.tree_flatten(input)
 
+    if reverse:
+        leaves = [torch.flip(elem, [dim]) for elem in leaves]
```
Comment (vadimkantorov) [inline torch/_higher_order_ops/associative_scan.py:88]: In practice `elem.flip(dim)` also works but unfortunately this is not documented properly:
- https://github.com/pytorch/pytorch/issues/68616

File: test/functorch/test_control_flow.py [hunk 2]
```diff
@@ -1174,6 +1203,225 @@ def fwbw(map_op, f, x, y):
         fake_outs = fwbw(_fake_map, f, x, y)
         self.assertEqual(true_outs, fake_outs)
 
+    @unittest.skipIf(not torch.cuda.is_available(), "Test requires CUDA.")
+    def test_generic_associative_scan_simple(self):
```
Comment (ydwu4) [inline test/functorch/test_control_flow.py:1207]: Nit: change test name to test_pointwise_associative_scan_reverse_simple?

File: test/functorch/test_control_flow.py [hunk 3]
```diff
@@ -1174,6 +1203,225 @@ def fwbw(map_op, f, x, y):
...
+        cumsum_exp = _fake_associative_scan(add, x, 0)
+        self.assertEqual(cumsum1, torch.tensor([0.0, 1.0, 3.0, 6.0], dtype=torch.int64))
+        self.assertEqual(cumsum1, cumsum_exp)
+
+        cumsum1 = associative_scan(add, x, 0, reverse=True)
+        cumsum_exp = _fake_associative_scan(add, x, 0, reverse=True)
+        self.assertEqual(cumsum1, torch.tensor([6.0, 6.0, 5.0, 3.0], dtype=torch.int64))
+        self.assertEqual(cumsum1, cumsum_exp)
+
+        num_dims = [random.randint(2, 5) for _ in range(10)]
```
Comment (ydwu4) [inline test/functorch/test_control_flow.py:1253]: nit: maybe a separate test for testing the dim?

File: test/functorch/test_control_flow.py [hunk 4]
```diff
@@ -1174,6 +1203,225 @@ def fwbw(map_op, f, x, y):
         fake_outs = fwbw(_fake_map, f, x, y)
         self.assertEqual(true_outs, fake_outs)
 
+    @unittest.skipIf(not torch.cuda.is_available(), "Test requires CUDA.")
```
Comment (ydwu4) [inline test/functorch/test_control_flow.py:1206]: We could add use the `parametrize` decorator for reverse=True/False when applicable.

File: test/functorch/test_control_flow.py [hunk 5]
```diff
@@ -1174,6 +1203,225 @@ def fwbw(map_op, f, x, y):
...
+    def test_generic_associative_scan_compile(self):
+        def add(x: torch.Tensor, y: torch.Tensor):
+            return x + y
+
+        def mul(x: torch.Tensor, y: torch.Tensor):
+            return x * y
+
+        x = torch.randn(3, 10, 2, device=torch.device("cuda"))
+        torch.compiler.reset()
+        with torch._dynamo.utils.disable_cache_limit():
```
Comment (ydwu4) [inline test/functorch/test_control_flow.py:1276]: why is this necessary? I thought we already disabled cache limit in https://github.com/pytorch/pytorch/pull/133011/files#diff-43442eb487890d221156357e289c9ba39d7ee3da0a96bd7b13448b4716a4cfcaL78

File: test/functorch/test_control_flow.py [hunk 6]
```diff
@@ -1174,6 +1203,225 @@ def fwbw(map_op, f, x, y):
...
+            associative_scan1 = torch.compile(
+                associative_scan, fullgraph=True, dynamic=False
+            )
+            associative_scan2 = torch.compile(
+                associative_scan, fullgraph=True, dynamic=True
+            )
+            associative_scan3 = associative_scan
```
Comment (ydwu4) [inline test/functorch/test_control_flow.py:1283]: We could parametrize reverse, dynamic like
```
@parametrize("reverse", [True, False])
@parametrize("compile_mode", ["eager", "compile", "compile_dynamic_shape"])
def test_generic_associative_scan_compile(self, reverse, compile_mode):
  if compile_mode == "eager":
    scan_fn = associative_scan
  elif compile_mode == "compile":
    scan_fn = torch.compile(
                    associative_scan, fullgraph=True, dynamic=False
      )
   else:
....
```

File: test/functorch/test_control_flow.py [hunk 7]
```diff
@@ -1174,6 +1203,225 @@ def fwbw(map_op, f, x, y):
...
+        for reverse in [False, True]:
+            torch.compiler.reset()
+            with torch._dynamo.utils.disable_cache_limit():
+                associative_scan1 = torch.compile(
+                    associative_scan, fullgraph=True, dynamic=False
+                )
+                associative_scan2 = torch.compile(
+                    associative_scan, fullgraph=True, dynamic=True
+                )
+                associative_scan3 = associative_scan
```
Comment (ydwu4) [inline test/functorch/test_control_flow.py:1294]: Redundant code?

File: test/functorch/test_control_flow.py [hunk 8]
```diff
@@ -1174,6 +1203,225 @@ def fwbw(map_op, f, x, y):
...
+        self.assertEqual(cumsum3, cumsum_exp)
+
+        cumsum1 = associative_scan1(add, x, 0, reverse=True)
+        cumsum2 = associative_scan2(add, x, 0, reverse=True)
+        cumsum3 = associative_scan2(add, x, 0, reverse=True)
+        cumsum_exp = _fake_associative_scan(add, x, 0, reverse=True)
+        self.assertEqual(cumsum1, torch.tensor([6.0, 6.0, 5.0, 3.0], dtype=torch.int64))
+        self.assertEqual(cumsum1, cumsum_exp)
+        self.assertEqual(cumsum2, cumsum_exp)
+        self.assertEqual(cumsum3, cumsum_exp)
```
Comment (ydwu4) [inline test/functorch/test_control_flow.py:1327]: We probably could delete these tests unless you think it covers cases that're different from the above

File: test/functorch/test_control_flow.py [hunk 9]
```diff
@@ -1174,6 +1203,225 @@ def fwbw(map_op, f, x, y):
...
+
+        x = torch.randn(3, 2, 2, device=torch.device("cuda"))
+
+        def f(op, x, dim, reverse):
+            result = associative_scan(op, x, dim, reverse)
+            return result
```
Comment (ydwu4) [inline test/functorch/test_control_flow.py:1333]: Delete this?

File: test/functorch/test_control_flow.py [hunk 10]
```diff
@@ -1174,6 +1203,225 @@ def fwbw(map_op, f, x, y):
...
+
+    @unittest.skipIf(not torch.cuda.is_available(), "Test requires CUDA.")
+    def test_generic_associative_scan_CUDA_flip(self):
```
Comment (ydwu4) [inline test/functorch/test_control_flow.py:1336]: change the name to pointwise scan?

File: test/functorch/test_control_flow.py [hunk 11]
```diff
@@ -1174,6 +1203,225 @@ def fwbw(map_op, f, x, y):
...
+
+        def fct(x: torch.Tensor, y: torch.Tensor):
+            return x + y
+
+        # This specific length was failing in the past
+        fails_for_backend = []
+        n = 9
+        x = torch.arange(n, device=device)
+        # for backend in ['eager', 'aot_eager', 'aot_eager_decomp_partition', 'inductor']:
+        for backend in ["inductor"]:
```
Comment (ydwu4) [inline test/functorch/test_control_flow.py:1347]: Can move inductor tests in test/inductor/test_control_flow.py. Can ignore other backends if inductor backend passes.

File: test/functorch/test_control_flow.py [hunk 12]
```diff
@@ -1174,6 +1203,225 @@ def fwbw(map_op, f, x, y):
...
+                associative_scan1 = torch.compile(
+                    associative_scan, backend=backend, fullgraph=True
+                )
+                associative_scan2 = associative_scan
+
+            try:
+                # Flip only non-compiled and compare with compiled reverse=True
+                result1 = associative_scan1(fct, x, 0, reverse=True)
+            except torch._dynamo.debug_utils.AccuracyError:
+                fails_for_backend.append(backend)
```
Comment (ydwu4) [inline test/functorch/test_control_flow.py:1359]: Why we have a try except?

File: test/functorch/test_control_flow.py [hunk 13]
```diff
@@ -1174,6 +1203,225 @@ def fwbw(map_op, f, x, y):
...
+            )
+            result3 = torch.flip(torch.cumsum(torch.flip(x, [0]), 0), [0])
+
+            self.assertEqual(result1, result2)
+            self.assertEqual(result1, result3)
+
+        print("Flip test fails for backends: " + str(fails_for_backend))
+        self.assertEqual(len(fails_for_backend), 0)
+
+        for n in range(20):
```
Comment (ydwu4) [inline test/functorch/test_control_flow.py:1371]: 20 might be too large? Don't want the unit tests take too long

File: test/functorch/test_control_flow.py [hunk 14]
```diff
@@ -1174,6 +1203,225 @@ def fwbw(map_op, f, x, y):
...
+                if not reverse:
+                    cumsum_exp_PT = op_pt(x, 0)
+                    self.assertEqual(cumsum1, cumsum_exp_PT)
+                else:
+                    # Check whether the graph contains at least two flip operations
+                    def f(op, inp, dim):
+                        result = associative_scan(op, inp, dim)
+                        return result
+
+                    # gm = make_fx(f, tracing_mode="symbolic")(op, x, 0)
```
Comment (ydwu4) [inline test/functorch/test_control_flow.py:1231]: This is because associative_scan is calling torch.compile with default "inductor" backend, we don't expect make_fx this compiled fn gives us the graph.

Can you try torch.compile(f, backend=EagerAndRecordGraphs())? Can follow `test_cond_simple_with_linear_compile_check_graph`.

**Review by ydwu4 — COMMENTED** [review]
Overall looks good to me. Left some comments for test organizations. We generally don't want unit test to be too complicated or take too long to run and we should split a large test into different tests that test different things.

**Comment by ydwu4** [comment]
Can skip the failed test with `from torch.testing._internal.common_cuda import SM70OrLater` unittest.skipIf(not SM70OrLater).

File: test/functorch/test_control_flow.py (hunk 14 above)
Comment (ydwu4) [reply test/functorch/test_control_flow.py:1231]: Forget about this. This doesn't work because pointwise associative scan didn't implement eager backend.

File: test/functorch/test_control_flow.py (hunk 2 above)
Comment (bohnstingl) [reply test/functorch/test_control_flow.py]: Done

File: test/functorch/test_control_flow.py (hunk 3 above)
Comment (bohnstingl) [reply test/functorch/test_control_flow.py:1270]: Done

File: test/functorch/test_control_flow.py (hunk 4 above)
Comment (bohnstingl) [reply test/functorch/test_control_flow.py:1206]: Done

File: test/functorch/test_control_flow.py (hunk 5 above)
Comment (bohnstingl) [reply test/functorch/test_control_flow.py]: I removed the `with torch._dynamo.utils.disable_cache_limit():` statements

File: test/functorch/test_control_flow.py (hunk 6 above)
Comment (bohnstingl) [reply test/functorch/test_control_flow.py]: I parameterized the test and moved it to `test/inductor/test_control_flow.py`

File: test/functorch/test_control_flow.py (hunk 11 above)
Comment (bohnstingl) [reply test/functorch/test_control_flow.py]: Done

File: test/functorch/test_control_flow.py (hunk 10 above)
Comment (bohnstingl) [reply test/functorch/test_control_flow.py]: I renamed the tests

File: test/functorch/test_control_flow.py (hunk 12 above)
Comment (bohnstingl) [reply test/functorch/test_control_flow.py]: I removed the try-except block

File: test/functorch/test_control_flow.py (hunk 13 above)
Comment (bohnstingl) [reply test/functorch/test_control_flow.py]: Well, the problem is that with the combine_fn='generic' that will follow, the test will randomly fail at some point. Thus I wanted to keep it like this. However, if you feel that it takes too long, I can reduce.

Comment (ydwu4) [reply test/functorch/test_control_flow.py]: Do we know why it fails randomly?

**Review by ydwu4 — APPROVED** [review]
Looks good! Wait for CI.

File: test/functorch/test_control_flow.py (hunk 13 above)
Comment (bohnstingl) [reply test/functorch/test_control_flow.py]: This test fails for `n=9` for the case `combine_mode='generic'` . The output of
```
torch.compile(
    associative_scan, backend=backend, fullgraph=True
) 
```
operation is
`tensor([36, 36, 35, 33, 30, 26, 21, 15,  8], device='cuda:0')` which is correct, while the output of
```
associative_scan2 = associative_scan
```
is
`tensor([36, 15, 35,  0, 30,  0, 21,  0,  8], device='cuda:0')`, which is incorrect.

This problem only appears for n=9, while for other values of n, the outputs are correct.
